### PR TITLE
Add Chapter "Debugging with CCPP"

### PR DESCRIPTION
**Pull request by bluefinweiwei**
new file:   source/CCPPDebug.rst	
modified:   source/index.rst

### CODE REVIEW

**Comment by climbfuji** [comment]
Did you create the documentation with those changes? Is it available somewhere for me to look at it? cheyenne or hera would do. Thanks ...

**Comment by bluefinweiwei** [comment]
Yep. Here they are: /glade/u/home/weiweili/my_work/GMTB/ccpp-doc/CCPPtechnical/source

Thanks!

**Comment by bluefinweiwei** [comment]
> Did you create the documentation with those changes? Is it available somewhere for me to look at it? cheyenne or hera would do. Thanks ...

Yep. Here they are: /glade/u/home/weiweili/my_work/GMTB/ccpp-doc/CCPPtechnical/source

Thanks!

**Comment by climbfuji** [comment]
What I meant to ask was if you had created the `.html` and/or `.pdf` versions to look at, sorry.

**Comment by bluefinweiwei** [comment]
> What I meant to ask was if you had created the `.html` and/or `.pdf` versions to look at, sorry.

I built all these on my laptop. Now they should have been updated on Cheyenne. Please check /glade/u/home/weiweili/my_work/GMTB/ccpp-doc/CCPPtechnical/build/html

Thanks!

File: CCPPtechnical/source/CCPPDebug.rst [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+         ...
+         ! Coupling
+         call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Coupling%nirbmdi', Coupling%nirbmdi)
+         ...
+         ! Grid
+         call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Grid%xlon', Grid%xlon)
+         ...
+         end subroutine GFS_diagtoscreen_run
+
+    All output to ``stdout/stderr`` from this routine is prefixed with **'XXX: '** so that it can be easily removed from the log files using "grep -ve 'XXX: ' ..." if needed.
```
Comment (climbfuji) [inline CCPPtechnical/source/CCPPDebug.rst:80]: This sentence ended up in the code block in the html version - needs to be regular text.

**Review by climbfuji — CHANGES_REQUESTED** [review]

**Comment by ligiabernardet** [comment]
Can also be viwed at
https://weiwei-techdoc.readthedocs.io/en/latest/ParamSpecificOutput.html
Note: there is no repo hook. If Weiwei makes changes to her branch/PR, I
need to trigger a new build manually.

**Comment by climbfuji** [comment]
> Can also be viwed at https://weiwei-techdoc.readthedocs.io/en/latest/ParamSpecificOutput.html Note: there is no repo hook. If Weiwei makes changes to her branch/PR, I need to trigger a new build manually.
> […](#)

Thanks, this page doesn't have the new chapter 11 yet.

**Comment by ligiabernardet** [comment]
Try this: https://weiwei-techdoc.readthedocs.io/en/ccpp_debug/

File: CCPPtechnical/source/CCPPDebug.rst [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            write(0,'(a,2i5,1x,e16.7)')'YYY: i, blk, landfrac(i)   :', i, blkno, landfrac(i)
+            write(0,'(a,2i5,1x,e16.7)')'YYY: i, blk, lakefrac(i)   :', i, blkno, lakefrac(i)
+            write(0,'(a,2i5,1x,e16.7)')'YYY: i, blk, slmsk(i)      :', i, blkno, slmsk(i)   
+            write(0,'(a,2i5,1x,i5)')   'YYY: i, blk, islmsk(i)     :', i, blkno, islmsk(i)
+            !end if
+        end do
+
+
+-----------------------------------------------
+How to use these debugging schemes for the UFS?
```
Comment (ligiabernardet) [inline CCPPtechnical/source/CCPPDebug.rst:151]:
```suggestion
How to use these debugging schemes for the UFS
```

File: CCPPtechnical/source/CCPPDebug.rst [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    For example, the Unified Forecast System (UFS) models use blocked data structures, which - at present - requires different debugging schemes for the ``time_vary`` group in the SDF. The blocked data structures are commonly known as “GFS types”, are defined in ``GFS_typedefs.F90`` and exposed to the CCPP in ``GFS_typedefs.meta``. The rationale for this storage model is a better cache reuse by breaking up contiguous horizontal grid columns into N blocks with a predefined block size, and allocating each of the GFS types N times. For example, the 3-dimensional air temperature is stored as
```
Comment (ligiabernardet) [inline CCPPtechnical/source/CCPPDebug.rst:26]: UFS already defined in previous chapter
```suggestion
    For example, the UFS models use blocked data structures, which - at present - requires different debugging schemes for the ``time_vary`` group in the SDF. The blocked data structures are commonly known as “GFS types”, are defined in ``GFS_typedefs.F90`` and exposed to the CCPP in ``GFS_typedefs.meta``. The rationale for this storage model is a better cache reuse by breaking up contiguous horizontal grid columns into N blocks with a predefined block size, and allocating each of the GFS types N times. For example, the 3-dimensional air temperature is stored as
```

File: CCPPtechnical/source/CCPPDebug.rst [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+In order to debug code efficiently with CCPP, it is important to remember the conceptual differences between traditional, physics-driver based approaches and the ones with CCPP. 
+
+Traditional, physics-driver based approaches rely on hand-written physics drivers that connect the different physical parameterizations together and often contain a large amount of "glue code" required between the parameterizations. As such, the physics drivers usually have access to all variables that are used by the physical parameterizations, while individual parameterizations only have access to the variables that are passed in. Debugging either happens on the level of the physics driver or inside physical parameterizations. In both cases, print statements are inserted in one or more places (e.g. in the driver before/after parameterizations to debug). In the CCPP, there are no hand-written physics drivers. Instead, the physical parameterizations are glued together by a suite definition file (SDF) that lists the primary physical parameterizations and so-called interstitial parameterizations (containing the glue code, broken up into logical units) in the order of execution.
```
Comment (ligiabernardet) [inline CCPPtechnical/source/CCPPDebug.rst:13]: SDF already defined in previous chapter
```suggestion
Traditional, physics-driver based approaches rely on hand-written physics drivers that connect the different physical parameterizations together and often contain a large amount of "glue code" required between the parameterizations. As such, the physics drivers usually have access to all variables that are used by the physical parameterizations, while individual parameterizations only have access to the variables that are passed in. Debugging either happens on the level of the physics driver or inside physical parameterizations. In both cases, print statements are inserted in one or more places (e.g. in the driver before/after parameterizations to debug). In the CCPP, there are no hand-written physics drivers. Instead, the physical parameterizations are glued together by a SDF that lists the primary physical parameterizations and so-called interstitial parameterizations (containing the glue code, broken up into logical units) in the order of execution.
```

**Review by ligiabernardet — CHANGES_REQUESTED** [review]
Looks good, just need some minor fixes.

File: CCPPtechnical/source/CCPPDebug.rst (hunk 1 above)
Comment (bluefinweiwei) [reply CCPPtechnical/source/CCPPDebug.rst]: Thanks for catching that! It should have been fixed now.

**Comment by bluefinweiwei** [comment]
> Looks good, just need some minor fixes.

All should have been fixed. Let me know of anything else to be revised. Thank you both for the review! :)

**Review by ligiabernardet — APPROVED** [review]

**Review by climbfuji — APPROVED** [review]

**Comment by climbfuji** [comment]
@JulieSchramm do you want to take a look as well before I merge this? Thanks!

**Review by JulieSchramm — APPROVED** [review]
Looks good!

**Pull request merged**